lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+   *
+   * @param  {HttpMessage} message
+   * @param  {Buffer} data
+   * @param  {Function} callback
+   */
+  httpCompress(message, data, callback) {
+    if (message.headers['accept-encoding']) {
+      const encodings = message.headers['accept-encoding']
+        .split(',')
+        .map(e => e.trim().toLowerCase());
+
+      let algorithm;
+      let priority = -1;
+
+      for (const encoding of encodings) {
+        if (encoding.startsWith('gzip') || encoding.startsWith('deflate')) {
+          let [ _algorithm, _priority = 'q=0' ] = encoding.split(';');
+          _priority = Number.parseFloat(_priority.split('=')[1] || 0);
+
+          if (_priority > priority) {
+            algorithm = _algorithm;
+            priority = _priority;
+          }
+          // gzip should be the preferred algorithm in case of a tie
+          else if (Math.abs(priority - _priority) < 10e-3
+            && _algorithm === 'gzip'
+          ) {
+            algorithm = 'gzip';
+          }
+        }
+      }
+
+      if (algorithm === 'gzip') {
+        zlib.gzip(data, (err, compressed) => callback({
+          compressed: !err ? compressed : data,
+          encoding: !err ? 'gzip' : 'identity',
+        }));
+        return;
+      }
+      else if (algorithm === 'deflate') {
+        zlib.deflate(data, (err, compressed) => callback({
+          compressed: !err ? compressed : data,
+          encoding: !err ? 'deflate' : 'identity',
+        }));
+        return;
+      }
+    }
+
+    callback({
+      compressed: data,
+      encoding: 'identity',
+    });
+  }
+
+  httpUncompress (message, payload, cb) {
+    let encodings = message.headers['content-encoding'];
+    if (!encodings) {
+      cb(null, payload);
+      return;
+    }
+
+    encodings = encodings.split(',').map(e => e.trim().toLowerCase());
+
+    if (encodings.length > this.httpConfig.opts.maxEncodingLayers) {
+      cb(kerrorHTTP.get('too_many_encodings'));
+      return;
+    }
+
+    // encodings are listed in the same order they have been applied
+    // this means that we need to invert the list to correctly
+    // decode the message
+    encodings.reverse();
+
+    this.httpUncompressStep(encodings, payload, cb, 0);
+  }
+
+  httpUncompressStep (encodings, payload, cb, index) {
+    if (index === encodings.length) {
+      cb(null, payload);
+      return;
+    }
+
+    const encoding = encodings[index];
+
+    if (encoding === 'identity') {
+      this.httpUncompressStep(encodings, payload, cb, index + 1);
+    }
+    else if (encoding === 'gzip' || encoding === 'deflate') {
+      const fn = encoding === 'gzip' ? zlib.gunzip : zlib.inflate;
+
+      fn(payload, (err, inflated) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+
+        this.httpUncompressStep(encodings, inflated, cb, index + 1);
+      });
+    }
+    else {
+      cb(kerrorHTTP.get('unsupported_compression', encoding));
+    }
+  }
+
+  parseWebSocketOptions () {
```

Comment:
This shouldn't be the responsability of the config module?

Reply:
Yes, it should :+1:

Reply:
I've moved the option checks to the config module. However the "parseXxxOptions()" are still needed to prepare the actual protocols options according to what has been configured.